### PR TITLE
reenable ref emit for xml serializer in UAP

### PR DESCRIPTION
**Pull request by krwq**
Fixes: https://github.com/dotnet/corefx/issues/21292
Related to: https://github.com/dotnet/corefx/pull/20783

### CODE REVIEW

File: src/System.Private.Xml/src/System/Xml/Xslt/XslCompiledTransform.cs [hunk 1]
```diff
@@ -64,7 +64,7 @@ static XslCompiledTransform()
         private XmlWriterSettings _outputSettings = null;
         private QilExpression _qil = null;
 
-#if !uap
+#if !uap && !uapaot
```
Comment (danmosemsft) [inline src/System.Private.Xml/src/System/Xml/Xslt/XslCompiledTransform.cs]: Why are you retaining some of the #if uap/#if !uap's ?

Comment (krwq) [inline src/System.Private.Xml/src/System/Xml/Xslt/XslCompiledTransform.cs]: XSLT is disabled for both and serialization just for UAP

**Comment by zhenlan** [comment]
cc: @shmao for code review

File: src/System.Private.Xml/src/System.Private.Xml.csproj [hunk 2]
```diff
@@ -8,6 +8,7 @@
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <NoWarn>$(NoWarn),649</NoWarn>
     <DefineConstants Condition="'$(TargetGroup)' == 'uap'">$(DefineConstants);uap</DefineConstants>
```
Comment (shmao) [inline src/System.Private.Xml/src/System.Private.Xml.csproj]: Can we change 
```
<DefineConstants Condition="'$(TargetGroup)' == 'uap'">$(DefineConstants);uap</DefineConstants>
```
to 
```
<DefineConstants Condition="'$(TargetGroup)' == 'uap' OR '$(TargetGroup)' == 'uapaot'">$(DefineConstants);uap</DefineConstants>
```
so that we can simply use `#if uap` or `#if !uap` in the code?

File: src/System.Private.Xml/src/System/Xml/Xslt/XslCompiledTransform.cs (hunk 1 above)
Comment (shmao) [inline src/System.Private.Xml/src/System/Xml/Xslt/XslCompiledTransform.cs]: Just question: the reason to disable XSLT was RefEmit not available. But as serialization uses RefEmit on UAP, it means RefEmit works on UAP. If that's the case, do we still want to disable XSLT on UAP? Can we disable it on UAPAOT only?

Comment (jkotas) [inline src/System.Private.Xml/src/System/Xml/Xslt/XslCompiledTransform.cs]: > If that's the case, do we still want to disable XSLT on UAP?

Yes, because of we want to have same behavior between UAP and UAPAOT.

File: src/System.Private.Xml/src/System.Private.Xml.csproj (hunk 2 above)
Comment (jkotas) [inline src/System.Private.Xml/src/System.Private.Xml.csproj]: It would be useful to define `FEATURE_XXX` defines for this instead, so that we do not have to redo all ifdefs every time we change our mind about the exact config used.

File: src/System.Private.Xml/src/System.Private.Xml.csproj [hunk 3]
```diff
@@ -7,7 +7,9 @@
     <RootNamespace>System.Xml</RootNamespace>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <NoWarn>$(NoWarn),649</NoWarn>
-    <DefineConstants Condition="'$(TargetGroup)' == 'uap'">$(DefineConstants);uap</DefineConstants>
+    <FeatureCompiledXsl Condition="'$(TargetGroup)' != 'uap' AND '$(TargetGroup)' != 'uapaot'">true</FeatureCompiledXsl>
+    <DefineConstants Condition="'$(FeatureCompiledXsl)' == 'true'">$(DefineConstants);FEATURE_COMPILED_XSL</DefineConstants>
+    <DefineConstants Condition="'$(TargetGroup)' == 'uapaot'">$(DefineConstants);FEATURE_SERIALIZATION_UAPAOT</DefineConstants>
```
Comment (jkotas) [inline src/System.Private.Xml/src/System.Private.Xml.csproj]: This should not have UAPAOT in the name. The name should describe what it actually does, like `FEATURE_COMPILED_SERIALIZATION` or `FEATURE_REFLECTIONBASED_SERIALIZATION`

File: src/System.Private.Xml/tests/XmlSerializer/XmlSerializerTests.cs [hunk 4]
```diff
@@ -3279,7 +3279,7 @@ public static void Xml_HiddenDerivedFieldTest()
         Assert.Equal(value.value, ((DerivedClass)actual).value);
     }
 
-#if !uap
+#if !FEATURE_SERIALIZATION_UAPAOT
```
Comment (shmao) [inline src/System.Private.Xml/tests/XmlSerializer/XmlSerializerTests.cs]: This should be `uapaot`. This is for test.

File: src/System.Private.Xml/tests/XmlSerializer/ReflectionOnly/System.Xml.XmlSerializer.ReflectionOnly.Tests.csproj [hunk 5]
```diff
@@ -4,12 +4,12 @@
   <PropertyGroup>
     <DefineConstants>$(DefineConstants);ReflectionOnly</DefineConstants>
     <ProjectGuid>{4050F1D1-1DD2-4B48-A17B-E3F955518C4B}</ProjectGuid>
-    <DefineConstants Condition="'$(TargetGroup)' == 'uap'">$(DefineConstants);uap</DefineConstants>
+    <DefineConstants Condition="'$(TargetGroup)' == 'uapaot'">$(DefineConstants);FEATURE_SERIALIZATION_UAPAOT</DefineConstants>
```
Comment (shmao) [inline src/System.Private.Xml/tests/XmlSerializer/ReflectionOnly/System.Xml.XmlSerializer.ReflectionOnly.Tests.csproj]: I'd prefer to use `uapaot` for test, which would mean the test is for uapaot specific.

**Comment by krwq** [comment]
fixing (likely some typo)

File: src/System.Private.Xml/tests/XmlSerializer/ReflectionOnly/System.Xml.XmlSerializer.ReflectionOnly.Tests.csproj (hunk 5 above)
Comment (krwq) [inline src/System.Private.Xml/tests/XmlSerializer/ReflectionOnly/System.Xml.XmlSerializer.ReflectionOnly.Tests.csproj]: @shmao do we not want to be consistent with product with `#ifdef`s?

File: src/System.Private.Xml/src/System.Private.Xml.csproj (hunk 3 above)
Comment (shmao) [inline src/System.Private.Xml/src/System.Private.Xml.csproj]: How about `FEATURE_USE_AOT_SERIALIZER` ?

File: src/System.Private.Xml/tests/XmlSerializer/ReflectionOnly/System.Xml.XmlSerializer.ReflectionOnly.Tests.csproj (hunk 5 above)
Comment (shmao) [inline src/System.Private.Xml/tests/XmlSerializer/ReflectionOnly/System.Xml.XmlSerializer.ReflectionOnly.Tests.csproj]: Good point. Can you please also change the `<DefineConstants>` in System.Xml.XmlSerializer.Tests.csproj then? Thanks!

Comment (shmao) [inline src/System.Private.Xml/tests/XmlSerializer/ReflectionOnly/System.Xml.XmlSerializer.ReflectionOnly.Tests.csproj]: > Good point. Can you please also change the <DefineConstants> in System.Xml.XmlSerializer.Tests.csproj then? Thanks!

Sorry. I was wrong above. This test behavior is uapaot specific. It's not specific for `FEATURE_SERIALIZATION_UAPAOT`. We should change this to `uapaot`.

Comment (krwq) [inline src/System.Private.Xml/tests/XmlSerializer/ReflectionOnly/System.Xml.XmlSerializer.ReflectionOnly.Tests.csproj]: @shmao good catch, I must've missed this one

File: src/System.Private.Xml/src/System.Private.Xml.csproj (hunk 3 above)
Comment (krwq) [inline src/System.Private.Xml/src/System.Private.Xml.csproj]: @jkotas we tried coming up with better name but that was the best we could come up with (without spending too much time on this)

Comment (jkotas) [inline src/System.Private.Xml/src/System.Private.Xml.csproj]: I think that it is confusing. There is nothing fundamentally AOT about it. The `FEATURE_` names should not be capturing the policy (which serializers are compiled into UAP or UAPAOT configs).

Take a look at how it is done in https://github.com/dotnet/corefx/blob/master/src/System.Linq.Expressions/src/System.Linq.Expressions.csproj#L13. This one should be similar. Also, notice the policy - we chose to compile the Linq expression interpreter into the CoreCLR builds, even though it is not strictly necessary for functional correctness (it is compiled in to give people choice about the perf characteristics).

File: src/System.Private.Xml/tests/XmlSerializer/ReflectionOnly/System.Xml.XmlSerializer.ReflectionOnly.Tests.csproj (hunk 5 above)
Comment (krwq) [inline src/System.Private.Xml/tests/XmlSerializer/ReflectionOnly/System.Xml.XmlSerializer.ReflectionOnly.Tests.csproj]: talked offline with @shmao, reverting all changes to tests since those are unrelated ifdefs

File: src/System.Private.Xml/src/System.Private.Xml.csproj (hunk 3 above)
Comment (krwq) [inline src/System.Private.Xml/src/System.Private.Xml.csproj]: @jkotas I do not have much context on serialization code and I've given similar argument to @shmao when we talked about how to name this but from our conversation it seems that this implementation is simply AOT specific. I wouldn't spend too much time on trying to figure out better name since all owners know what it's for and we have achieved the purpose of the switch of having one centralize place to change logic.

Comment (shmao) [inline src/System.Private.Xml/src/System.Private.Xml.csproj]: I suggested @krwq to `FEATURE_SERIALIZATION_UAPAOT` at the beginning because the serialization code that was under `uapaot` was the code that knows how to use the serializer pre-generated by UWP toolchain. As UWP toolchain is used on UAPAOT only, so I suggested that name.

But I agreed with you that we shouldn't put `uap` or `uapaot` in the name. So I thought about using names like `FEATURE_USE_PREGENERATED_SERIALIZER`. The concern about use this name was that it might cause confusion between using the pre-generated serializer by UWP toolchain and using the pre-generated serializer by SGen.exe (not toolchain).

I think `FEATURE_USE_AOT_SERIALIZER` would fit here the best, which means using the ahead of time serializer, i.e. generated by UWP toolchain.

Comment (jkotas) [inline src/System.Private.Xml/src/System.Private.Xml.csproj]: > we tried coming up with better name

`FEATURE_COMPILED_SERIALIZATION` - to make it similar to what you have done for XSL?

Comment (jkotas) [inline src/System.Private.Xml/src/System.Private.Xml.csproj]: >  `FEATURE_USE_AOT_SERIALIZER` would fit here the best,

I agree that the code pre-generated by UWP toolchain is pretty uapaot specific. We may not even need `FEATURE_XXX` for that one - just use uapaot.

**Comment by krwq** [comment]
@dotnet-bot test Linux x64 Release Build please

**Comment by krwq** [comment]
@dotnet-bot test Windows x86 Release Build please

**Comment by krwq** [comment]
Talked with @shmao offline and we will leave the serialization specific UAPAOT define.
